src/main/java/bio/terra/app/controller/SearchApiController.java

```diff
@@ -94,7 +98,23 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
         @Valid @RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
         @Valid @RequestParam(value = "limit", required = false, defaultValue = "1000") Integer limit
     ) {
-        SearchQueryResultModel searchQueryResultModel = new SearchQueryResultModel();
+        List<String> accessibleIds =
+                iamService.listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT)
+                        .stream()
+                        .map(UUID::toString)
+                        .collect(Collectors.toList());
+
+        List<String> requestIds = searchQueryRequest.getSnapshotIds();
+
+        for (String snapShotId : requestIds) {
+            if (!accessibleIds.contains(snapShotId)){
+                throw new IamForbiddenException("User '" + getAuthenticatedInfo().getEmail()
+                        + "' does not have required action: " + IamAction.READ_DATA + " on snapshot id" + snapShotId);
```

Comment:
As a general rule I prefer collecting all errors before reporting, so a user don't have to make multiple calls to find all errors. One way to express this is to use a set difference, e.g. `new HashSet<>(requestIds).removeAll(accessibleIds)`